### PR TITLE
Adds the `EquipmentDropChance` property

### PR DESCRIPTION
**Pull request by heypr**
- Requested by [LeafyMitsuwa](https://discord.com/channels/315163488085475337/1117428374328062056) on Discord.

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    // Specifying a drop chance of 0 will prevent the item from dropping, a drop chance of 1 will always drop the item if killed by a player, and a drop chance of higher than 1 will always drop the item no matter what the entity was killed by.
+    //
+    // @Usage
+    // Use to prevent a zombie from dropping any of its equipped items, no matter what:
+    // - adjust <[zombie]> equipment_drop_chance:[helmet=0;chestplate=0;leggings=0;boots=0;main_hand=0;off_hand=0]
+    //
+    // -->
+
+    public static boolean describes(EntityTag entity) {
+        return true;
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: I believe this is living entity only?

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    @Override
+    public MapTag getPropertyValue() {
+        EntityEquipment equipment = getLivingEntity().getEquipment();
+        if (equipment == null) {
+            return null;
+        }
+        MapTag map = new MapTag();
+        if (equipment.getHelmet() != null) {
+            map.putObject("helmet", new ElementTag(equipment.getHelmetDropChance()));
+        }
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: Any specific reason to exclude the drop chance if the item isn't present? seeing as it's `float` (not nullable) & looking at internals, it seems to be available regardless of whether there's an item or not

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java (hunk 1 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: Actually looking at the Javadocs this seems to be limited to `Mob`s
image

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    // <--[property]
+    // @object EntityTag
+    // @name equipment_drop_chance
+    // @input MapTag
+    // @description
+    // Controls the chance of each piece of equipment dropping when the entity dies.
+    // Specifying a drop chance of 0 will prevent the item from dropping, a drop chance of 1 will always drop the item if killed by a player, and a drop chance of higher than 1 will always drop the item no matter what the entity was killed by.
+    //
+    // @Usage
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: `usage` meta is for commands, you want `example` meta

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    // <--[property]
+    // @object EntityTag
+    // @name equipment_drop_chance
+    // @input MapTag
+    // @description
+    // Controls the chance of each piece of equipment dropping when the entity dies.
+    // Specifying a drop chance of 0 will prevent the item from dropping, a drop chance of 1 will always drop the item if killed by a player, and a drop chance of higher than 1 will always drop the item no matter what the entity was killed by.
+    //
+    // @Example
+    // Use to prevent a zombie from dropping any of its equipped items, no matter what:
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: this is not proper example meta format

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        return map;
+    }
+
+    @Override
+    public void setPropertyValue(MapTag map, Mechanism mechanism) {
+        if(!mechanism.hasValue()) {
+            return;
+        }
+        EntityEquipment equipment = getLivingEntity().getEquipment();
+        if (equipment == null) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: Why is this null check here, and not the get? If it's needed it should be in both, if it's not it shouldn't be in either.

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        map.putObject("main_hand", new ElementTag(equipment.getItemInMainHandDropChance()));
+        map.putObject("off_hand", new ElementTag(equipment.getItemInOffHandDropChance()));
+        return map;
+    }
+
+    @Override
+    public void setPropertyValue(MapTag map, Mechanism mechanism) {
+        if(!mechanism.hasValue()) {
+            return;
+        }
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: Any reason for this check? `autoRegister` should already require input

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        EntityEquipment equipment = getLivingEntity().getEquipment();
+        ElementTag helmet = map.getElement("helmet");
+        ElementTag chestplate = map.getElement("chestplate");
+        ElementTag leggings = map.getElement("leggings");
+        ElementTag boots = map.getElement("boots");
+        ElementTag main_hand = map.getElement("main_hand");
+        ElementTag off_hand = map.getElement("off_hand");
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: Java naming conventions

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    // @object EntityTag
+    // @name equipment_drop_chance
+    // @input MapTag
+    // @description
+    // Controls the chance of each piece of equipment dropping when the entity dies.
+    // Specifying a drop chance of 0 will prevent the item from dropping, a drop chance of 1 will always drop the item if killed by a player, and a drop chance of higher than 1 will always drop the item no matter what the entity was killed by.
+    //
+    // @Example
+    // # Use to prevent a zombie from dropping any of its equipped items, no matter what:
+    // - adjust <[zombie]> equipment_drop_chance:[helmet=0;chestplate=0;leggings=0;boots=0;main_hand=0;off_hand=0]
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: `@mechanism-example` is a thing now, probably makes sense to use here seeing as this isn't relevant for the tag.
Or alternatively make it an example for both, I.e. setting it to `.drop_chance.with[leggings].as[0]` to make sure leggings don't drop while keeping other chances or something.

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+public class EntityEquipmentDropChance extends EntityProperty<MapTag> {
+
+    // <--[property]
+    // @object EntityTag
+    // @name equipment_drop_chance
+    // @input MapTag
+    // @description
+    // Controls the chance of each piece of equipment dropping when the entity dies.
+    // Specifying a drop chance of 0 will prevent the item from dropping, a drop chance of 1 will always drop the item if killed by a player, and a drop chance of higher than 1 will always drop the item no matter what the entity was killed by.
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: Should specify what exactly _is_ the map here (e.g. map of equipment slots to drop chances).
Also nitpick but I'd remove `Specifying` to make it a bit nicer for the tag.

**Review by mcmonkey4eva — DISMISSED** [review]

**Comment by mcmonkey4eva** [comment]
heh, this has a merge conflict with your own previous PR

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    @Override
+    public MapTag getPropertyValue() {
+        EntityEquipment equipment = getLivingEntity().getEquipment();
+        MapTag map = new MapTag();
+        map.putObject("helmet", new ElementTag(equipment.getHelmetDropChance()));
+        map.putObject("chestplate", new ElementTag(equipment.getChestplateDropChance()));
+        map.putObject("leggings", new ElementTag(equipment.getLeggingsDropChance()));
+        map.putObject("boots", new ElementTag(equipment.getBootsDropChance()));
+        map.putObject("main_hand", new ElementTag(equipment.getItemInMainHandDropChance()));
+        map.putObject("off_hand", new ElementTag(equipment.getItemInOffHandDropChance()));
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: Key names should probably match `EquipmentSlot` names for easier usage with other features

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    // @object EntityTag
+    // @name equipment_drop_chance
+    // @input MapTag
+    // @description
+    // Controls the chance of each piece of equipment dropping when the entity dies.
+    // A drop chance of 0 will prevent the item from dropping, a drop chance of 1 will always drop the item if killed by a player, and a drop chance of higher than 1 will always drop the item no matter what the entity was killed by.
+    //
+    // @Mechanism-example
+    // # Use to prevent a zombie from dropping any of its equipped items, no matter what:
+    // - adjust <[zombie]> equipment_drop_chance:[helmet=0;chestplate=0;leggings=0;boots=0;main_hand=0;off_hand=0]
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityEquipmentDropChance.java]: Key names are outdated (also `Mechanism` should be a lowercase `m`)

**Review by mcmonkey4eva — DISMISSED** [review]

**Review by mcmonkey4eva — APPROVED** [review]

**Review by tal5 — APPROVED** [review]

**Pull request merged**